tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc

```diff
@@ -629,26 +653,103 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
     HloInstruction *c_bcast = instr->AddInstruction(
         HloInstruction::CreateBroadcast(instr->shape(), c, {}));
 
+    // Each operand must have exactly one contracting and one non-contracting
+    // dimension.
+    absl::Span<const int64_t> a_contracting_dims =
+        gemm_backend_config.dot_dimension_numbers()
+            .lhs_contracting_dimensions();
+    absl::Span<const int64_t> b_contracting_dims =
+        gemm_backend_config.dot_dimension_numbers()
+            .rhs_contracting_dimensions();
+    if (a_contracting_dims.size() != 1 || b_contracting_dims.size() != 1) {
+      return false;
+    }
+    if ((a_bitcast ? a_bitcast : a)->shape().dimensions_size() -
+                gemm_backend_config.dot_dimension_numbers()
+                    .lhs_batch_dimensions()
+                    .size() !=
+            2 ||
+        (b_bitcast ? b_bitcast : b)->shape().dimensions_size() -
+                gemm_backend_config.dot_dimension_numbers()
+                    .rhs_batch_dimensions()
+                    .size() !=
+            2) {
+      return false;
+    }
+
+    // Shift any bitcasts to the unconverted and unscaled operands.
+    if (a_bitcast) {
+      a = instr->AddInstruction(a_bitcast->CloneWithNewOperands(
+          ShapeUtil::MakeShape(a->shape().element_type(),
+                               a_bitcast->shape().dimensions()),
+          {a}));
+    }
+    if (b_bitcast) {
+      b = instr->AddInstruction(b_bitcast->CloneWithNewOperands(
+          ShapeUtil::MakeShape(b->shape().element_type(),
+                               b_bitcast->shape().dimensions()),
+          {b}));
+    }
+
     // cuBLASLt FP8 GEMM kernels currently require the first operand, i.e. A, to
     // be transposed. If the result of the GEMM is not in column major order, A
     // and B are later exchanged, and B is transposed here instead.
     // TODO(philipphack): Remove once cuBLASLt supports the NN configuration.
-    TF_ASSIGN_OR_RETURN(auto gemm_config,
-                        instr->backend_config<GemmBackendConfig>());
     TF_ASSIGN_OR_RETURN(bool is_col_major,
-                        OutputIsColumnMajor(instr, gemm_config));
+                        OutputIsColumnMajor(instr, gemm_backend_config));
+
+    // Identify the dimensional order which describes a transpose of the
+    // contracting and non-contracting dimensions of the GEMM.
+    auto transp_dim_order =
+        [](HloInstruction *x, absl::Span<const int64_t> x_contracting_dims,
```

Comment:
The logic would be clearer if you passed just the single contracting dim as an int